src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/AutoInsert/AutoInsertService.cs

```diff
@@ -0,0 +1,85 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Frozen;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.PooledObjects;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.Razor.AutoInsert;
+
+internal class AutoInsertService(IEnumerable<IOnAutoInsertProvider> onAutoInsertProviders) : IAutoInsertService
+{
+    private readonly ImmutableArray<IOnAutoInsertProvider> _onAutoInsertProviders = onAutoInsertProviders.ToImmutableArray();
+
+    public static FrozenSet<string> HtmlAllowedAutoInsertTriggerCharacters { get; }
+        = new string[] { "=" }.ToFrozenSet(StringComparer.Ordinal);
+    public static FrozenSet<string> CSharpAllowedAutoInsertTriggerCharacters { get; }
+        = new string[] { "'", "/", "\n" }.ToFrozenSet(StringComparer.Ordinal);
+
+    private readonly ImmutableArray<string> _triggerCharacters = CalculateTriggerCharacters(onAutoInsertProviders);
+
+    private static ImmutableArray<string> CalculateTriggerCharacters(IEnumerable<IOnAutoInsertProvider> onAutoInsertProviders)
+    {
+        var builder = ImmutableArray.CreateBuilder<string>(onAutoInsertProviders.Count());
```

Comment:
It doesn't have Cointains on it I was using to de-dupe. I can switch to .Any, not sure there are concerns about iterators there. I'm not sure if it's worth it to de-dupe, especially considering there are only two providers now and we aren't anticipating any more. I'll switch to Any for now, please let me know if you have other suggestions.

Also FWIW it's a 2-element immutable array here. Per your suggestion we may convert this to inline creation instead of DI injection, so I'm not sure it's worth spending a lot of time this.

Reply:
Switched to using PooledArrayBuilder